lib/guest_installaltion_and_configuration_base.pm

```diff
@@ -0,0 +1,1428 @@
+        }
+        else {
+            $self->do_attach_guest_installation_screen;
+        }
+    }
+    else {
+        record_info("Guest $self->{guest_name} installation screen process $self->{guest_installation_session} had already been attached", "Good news !");
+    }
+    return $self;
+}
+
+#Call do_attach_guest_installation_screen_with_session anyway.Mark [guest_installation_attached] as true if needle 'text-logged-in-root' can not be detected.
+#If fails to attach guest installation screen, [guest_installation_session] may terminate at reboot/shutoff or be in mysterious state or just broken somehow,
+#call do_attach_guest_installation_screen_without_sesssion to re-attach.
+sub do_attach_guest_installation_screen {
+    my $self = shift;
+
+    $self->reveal_myself;
+    $self->do_attach_guest_installation_screen_with_session;
+    if (!(check_screen('text-logged-in-root'))) {
+        $self->{guest_installation_attached} = 'true';
+        record_info("Attached $self->{guest_name} installation screen process $self->{guest_installation_session} successfully", "Well done !");
+    }
+    else {
+        if ($self->{guest_autoconsole} eq '') {
+            record_info("Guest $self->{guest_name} has no autoconsole and installation screen process $self->{guest_installation_session} may terminate on reboot/shutoff after installaton finishes or at certain stage", "Reconnect by using screen -t $self->{guest_name} virsh console $self->{guest_name}");
+        }
+        else {
+            record_info("Failed to attach $self->{guest_name} installation screen process $self->{guest_installation_session}", "Bad luck ! Try to re-connect by using screen -t $self->{guest_name} virsh console $self->{guest_name}");
+        }
+        $self->do_attach_guest_installation_screen_without_session;
+    }
+    return $self;
+}
+
+#Retry attach [guest_installation_session] and detect needle 'text-logged-in-root'.
+sub do_attach_guest_installation_screen_with_session {
+    my $self = shift;
+
+    $self->reveal_myself;
+    assert_screen('text-logged-in-root');
+    type_string("reset\n");
+    save_screenshot;
+    my $_retry_counter = 3;
+    while (check_screen('text-logged-in-root', timeout => 5)) {
+        if ($_retry_counter gt 0) {
+            wait_screen_change {
+                type_string("screen -d -r $self->{guest_installation_session}\n");
+            };
+            save_screenshot;
+            $_retry_counter--;
+        }
+        else {
+            save_screenshot;
+            last;
+        }
+        save_screenshot;
+    }
+    save_screenshot;
+    return $self;
+}
+
+#If [guest_installation_session] is already terminated at reboot/shutoff or somehow, power it on and retry attaching using [guest_installation_session_command] and detect
+#needle 'text-logged-in-root'.Mark it as FAILED if needle 'text-logged-in-root' can still be detected and poweron can not bring it back.
+sub do_attach_guest_installation_screen_without_session {
```

Comment:
Firstly, do_attach_guest_installation_screen_with_session retry attach [guest_installation_session].
Secondly, do_attach_guest_installation_screen_without_session just power on guest instead of reset or reboot.